### PR TITLE
ci: add the support for release pipeline

### PR DESCRIPTION
**Pull request by apolikamixitos**
# What:

- Added support for the release pipeline for the different packages of the project.

# Description:

### `create-release-pr` workflow:

- Creates a release PR using `release-plz` tool based on the git history commits and generates the corresponding `changelog` file changes for each package separately (as discussed and confirmed with the team).

image

### `release` workflow:

- This workflow is triggered once the `release PR` gets merged and does the following:
    - Publishes the crates to the registry
    - Creates the corresponding Git releases and tags:
       Screenshot 2024-11-05 at 12 21 20 PM

    - Build the optimized artifacts of the project (`wasm` and `archive` files to be specific) and attach them to:
        -  The GitHub releases:
        Screenshot 2024-11-05 at 12 21 20 PM

        - The secondary artifacts distribution channel `R2`:
    
        image

### CODE REVIEW

File: .github/workflows/create-release-pr.yaml [hunk 1]
```diff
@@ -0,0 +1,54 @@
+# This workflow creates the PR that will trigger the new release and update the versions of the packages.
+# It's triggered manually once the team is ready to push a new release.
+name: Create Release PR
+
+on:
+  workflow_dispatch:
+
+concurrency: ${{ github.workflow }}-${{ github.ref }}
```
Comment (RiceAndMeet) [inline .github/workflows/create-release-pr.yaml:8]: Is this only allowing one instance of job running concurrently per commit?

File: .github/workflows/create-release-pr.yaml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+jobs:
+  create-release-pr:
+    name: Create Release PR
+    runs-on: blacksmith-2vcpu-ubuntu-2204
+
+    steps:
+      - name: Check if a branch is whitelisted and maintained
```
[the remark this answers is not in the document]
Comment (RiceAndMeet) [reply .github/workflows/create-release-pr.yaml:16]: This is hardcoded inline `WHITELISTED_BRANCHES=("main" "releases/*")` 

Correct me if am wrong @milapsheth Based on what I understood, we won't be using release branches at all, everything PR will be converge and  merged into main directly. Maybe we can just whitelist `main` branch here.

File: .github/workflows/release.yaml [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+          fetch-depth: 0
+          token: ${{ secrets.PAT_TOKEN }}
+
+      - name: Install Rust toolchain
+        uses: dtolnay/rust-toolchain@stable
+  
+      # Creates git tags and publishes the crates of the new releases
+      - name: Publish release
+        id: publish-release
+        uses: MarcoIeni/release-plz-action@fff938ea2923fdfa35797fff63537b00f30fe6f9 # v0.5.82
```
[the remark this answers is not in the document]
Comment (RiceAndMeet) [reply .github/workflows/release.yaml:45]: I see, it did mention only use tagged version only if we can trust the creator, even then a tag can be altered 
> if a bad actor gains access to the repository storing the action.  

I think its kinda overkill but I guess never hurts to be too careful, just a bit inconvenient to read.

File: .github/workflows/create-release-pr.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (RiceAndMeet) [reply .github/workflows/create-release-pr.yaml:16]: Answering my own question. in the case of orphan commit for patch fixes, I guess we're going to use releases/* branches.

File: .github/workflows/reusable-build-upload.yaml [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        uses: dtolnay/rust-toolchain@stable
+        with:
+          toolchain: 1.76.0
+          targets: wasm32-unknown-unknown
+
+      - name: Build artifacts for ${{ inputs.package-name }}-v${{ inputs.package-version }}
+        run: |
+          echo "Building wasm for '${{ inputs.package-name }}-v${{ inputs.package-version }}'";
+          # 21.1.1 is the latest rust 1.76.0 compatible version, the version after is 21.2.0 which is compatible with 1.79.0 (different than the workspace's version)
+          cargo install --locked soroban-cli --version 21.1.1 --features opt
```
Comment (ahramy) [inline .github/workflows/reusable-build-upload.yaml:70]: the Stellar Testnet upgraded to v22 recently, the version should be updated accordingly. ref: https://github.com/axelarnetwork/axelar-cgp-soroban/pull/64

Comment (RiceAndMeet) [reply .github/workflows/reusable-build-upload.yaml:70]: I think the release candidate will be released on Nov 17, we can open another pr then and shift off the "rc" version.

**Pull request merged**